overlord/snapstate/snapmgr.go

```diff
@@ -775,11 +829,20 @@ func (m *SnapManager) undoLinkSnap(t *state.Task, _ *tomb.Tomb) error {
 	if err != nil {
 		return err
 	}
-
+	var hadCandidate bool
+	err = t.Get("had-candidate", &hadCandidate)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
 	// relinking of the old snap is done in the undo of unlink-current-snap
-
-	snapst.Candidate = snapst.Sequence[len(snapst.Sequence)-1]
-	snapst.Sequence = snapst.Sequence[:len(snapst.Sequence)-1]
+	snapst.Candidate = snapst.Sequence[snapst.findIndex(snapst.Current)]
+	if !hadCandidate {
+		i := snapst.findIndex(snapst.Current)
+		if i < 0 {
+			return fmt.Errorf("internal error: cannot find revision %d in %v for undoing the added revision", snapst.Candidate.Revision, snapst.Sequence)
```

Comment:
This is unreachable. If i < 0, Sequence was indexed with a negative number just above, and so we had a panic.

The error checking is good, though. Let's just move this up and put the index on a var:
```
currentIndex := snapst.findIndex(snapst.Current)
if currentIndex < 0 { ... }
snapst.Candidate = ...
if !hadCandidate { ... }
```